tools/nite/main.cpp

```diff
@@ -0,0 +1,386 @@
+		{
+			m_logger->log("Lowering frames in flight!", ILogger::ELL_WARNING);
+			m_maxFramesInFlight = FRAMES_IN_FLIGHT;
+		}
+
+		m_cmdPool = m_device->createCommandPool(gQueue->getFamilyIndex(), IGPUCommandPool::CREATE_FLAGS::RESET_COMMAND_BUFFER_BIT);
+
+		for (auto i = 0u; i < m_maxFramesInFlight; i++)
+		{
+			if (!m_cmdPool)
+				return logFail("Couldn't create Command Pool!");
+			if (!m_cmdPool->createCommandBuffers(IGPUCommandPool::BUFFER_LEVEL::PRIMARY, { m_cmdBufs.data() + i, 1 }))
+				return logFail("Couldn't create Command Buffer!");
+		}
+
+		ui = core::make_smart_refctd_ptr<nbl::ext::imgui::UI>(smart_refctd_ptr(m_device), (int)m_maxFramesInFlight, renderpass, nullptr, smart_refctd_ptr(m_window));
+		auto* ctx = reinterpret_cast<ImGuiContext*>(ui->getContext());
+		ImGui::SetCurrentContext(ctx);
+	
+		// Initialize Test Engine
+		engine = ImGuiTestEngine_CreateContext();
+		ImGuiTestEngineIO& test_io = ImGuiTestEngine_GetIO(engine);
+		test_io.ConfigVerboseLevel = ImGuiTestVerboseLevel_Info;
+		test_io.ConfigVerboseLevelOnError = ImGuiTestVerboseLevel_Debug;
+		test_io.ConfigSavedSettings = false;
+
+		// Register tests
+		RegisterTests_All(engine);
+
+		// Start engine
+		ImGuiTestEngine_Start(engine, ctx);
+		ImGuiTestEngine_InstallDefaultCrashHandler();
+
+		ui->registerListener([this]() -> void
+			{
+				ImGuiTestEngine_ShowTestEngineWindows(engine, nullptr);
+			}
+		);
+
+		m_winMgr->setWindowSize(m_window.get(), WIN_W, WIN_H);
+		m_surface->recreateSwapchain();
+		m_winMgr->show(m_window.get());
+
+		return true;
+	}
+
+	inline void workLoopBody() override
+	{
+		static std::chrono::microseconds previousEventTimestamp{};
+		// TODO: Use real deltaTime instead
+		float deltaTimeInSec = 0.1f;
+
+		struct
+		{
+			std::vector<SMouseEvent> mouse{};
+		} capturedEvents;
+
+		m_inputSystem->getDefaultMouse(&mouse);
+		m_inputSystem->getDefaultKeyboard(&keyboard);
+
+		mouse.consumeEvents([&](const IMouseEventChannel::range_t& events) -> void
+			{
+				for (auto event : events)
+				{
+					if (event.timeStamp < previousEventTimestamp)
+						continue;
+
+					previousEventTimestamp = event.timeStamp;
+					capturedEvents.mouse.push_back(event);
+				}
+			}, m_logger.get());
+
+		keyboard.consumeEvents([&](const IKeyboardEventChannel::range_t& events) -> void
+			{
+				// TOOD
+			}, m_logger.get());
+
+		const auto mousePosition = m_window->getCursorControl()->getPosition();
+		ui->update(deltaTimeInSec, static_cast<float>(mousePosition.x), static_cast<float>(mousePosition.y), capturedEvents.mouse.size(), capturedEvents.mouse.data());
+
+		const auto resourceIx = m_realFrameIx % m_maxFramesInFlight;
+
+		if (m_realFrameIx >= m_maxFramesInFlight)
+		{
+			const ISemaphore::SWaitInfo cbDonePending[] =
+			{
+				{
+					.semaphore = m_semaphore.get(),
+					.value = m_realFrameIx + 1 - m_maxFramesInFlight
+				}
+			};
+			if (m_device->blockForSemaphores(cbDonePending) != ISemaphore::WAIT_RESULT::SUCCESS)
+				return;
+		}
```

Comment:
I think some surgery between two examples went wrong.

The new examples use `m_submitIx` to advance the semaphore each time the main render submission is done.

and here you're doing some sync based on `m_realdFrameIx` instead

Reply:
we agreed to fix/update/adjust the nite in the next PR